Provide consolidated averaging methods

brianpm opened the pull request:
I added these functions to `plotting_functions.py`:
- spatial_average : transplanted this from somewhere `amwg_table.py` to start consolidating all the averaging into one place
- annual_mean : provides a reliable way to produce time-weighted annual means
- seasonal_mean : provides a method to make seasonal means that are time-weighted

I then modified:
- `global_latlon_map.py`
- `zonal_mean.py`
- `meridional_mean.py`

To all use the same averaging (i.e., `seasonal_mean()`). 

I did **limited** testing to see if this works. My hope is that this helps to reduce any inconsistencies in the final output. For example, this should provide time-weighted means in both the table and the maps.

I moved the checks for the output files to *before* the averaging. That should allow a speed up when re-running the ADF because the calculations can be skipped once existing plots are detected. 


### an aside
I'm not sure whether we want to reconsider the option `weight_season`? I think whenever we have monthly data, we should use a weighted average (even for weird calendars that don't have different lengths of months, I guess). The weighting could be dropped for high-frequency output, but maybe that shouldn't a user-facing option?

nusbaume commented on scripts/plotting/global_latlon_map.py:
```diff
@@ -336,6 +283,20 @@ def global_latlon_map(adfobj):
+
+                            if weight_season:
+                                mseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
+                                oseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
```
I think we'll want to use `odata` here:

```suggestion
                                oseasons[s] = pf.seasonal_mean(odata, season=s, is_climo=True)
```

brianpm replied:
Well, that'd explain some of the strange behavior I saw in some recent testing.

nusbaume commented on scripts/plotting/global_latlon_map.py:
```diff
@@ -442,6 +375,21 @@ def global_latlon_map(adfobj):
+                                #If requested, then calculate the monthly-weighted seasonal averages:
+                                if weight_season:
+                                    mseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
+                                    oseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
```
I think we'll want to use `odata` here:

```suggestion
                                    oseasons[s] = pf.seasonal_mean(odata, season=s, is_climo=True)
```

nusbaume commented on scripts/plotting/global_latlon_map.py:
```diff
@@ -442,6 +375,21 @@ def global_latlon_map(adfobj):
+                                    mseasons[s] = mdata.sel(time=seasons[s], lev=pres).mean(dim='time')
+                                    oseasons[s] = odata.sel(time=seasons[s], lev=pres).mean(dim='time')
+                                    # difference: each entry should be (lat, lon)
+                                    dseasons[s] = mseasons[s] - oseasons[s]
```
Should `dseasons` be outside the `else`  block?

brianpm replied:
Yes.

nusbaume commented on scripts/plotting/meridional_mean.py:
```diff
@@ -220,32 +220,22 @@ def meridional_mean(adfobj):
                         plot_name.unlink()
 
+                    mseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
+                    oseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
```
I think we'll want to use `odata` here:

```suggestion
                    oseasons[s] = pf.seasonal_mean(odata, season=s, is_climo=True)
```

nusbaume commented on scripts/plotting/zonal_mean.py:
```diff
@@ -269,6 +252,16 @@ def zonal_mean(adfobj):
                     #End if
 
+                    #Seasonal Averages
+                    mseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True) #mdata.sel(time=seasons[s]).mean(dim='time')
+                    oseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True) #odata.sel(time=seasons[s]).mean(dim='time')
```
I think we'll want to use `odata` here, and get rid of the commented out code:

```suggestion
                    mseasons[s] = pf.seasonal_mean(mdata, season=s, is_climo=True)
                    oseasons[s] = pf.seasonal_mean(odata, season=s, is_climo=True)
```

nusbaume commented on scripts/plotting/meridional_mean.py:
```diff
@@ -220,32 +220,22 @@ def meridional_mean(adfobj):
-                    mseasons[s] = mdata.sel(time=seasons[s]).mean(dim='time')
-                    oseasons[s] = odata.sel(time=seasons[s]).mean(dim='time')
-
-                    # difference: each entry should be (lat, lon) or (plev, lat, lon)
-                    # dseasons[s] = mseasons[s] - oseasons[s]
-                    # difference will be calculated in plot_meridional_mean_and_save.
-
-                    # time to make plot; here we'd probably loop over whatever plots we want for this variable
-                    # I'll just call this one "Meridional_Mean"  ... would this work as a pattern [operation]_[AxesDescription] ?
-                    # NOTE: Up to this point, nothing really differs from global_latlon_map,
-                    #       so we could have made one script instead of two.
-                    #       Merging would make overall timing better because looping twice will double I/O steps.
-                    #
```
Just checking if you wanted to save this comment block down below, or completely get rid of it like what is being done currently?

brianpm replied:
I think I just wanted to delete it. I didn't think it was useful at this point.

nusbaume reviewed (changes requested):
Looks like a great improvement, thanks @brianpm!  I know it appears that I left a bunch of comments, but most of them should be trivial to resolve.

Also, I am happy to get rid of `weight_season` config argument.  I think the only reason I added it originally was because the old AMWG package had it and warned of a significant slowdown, but I don't get the sense that same slow down happens for the ADF.  Plus we are always adding new config variables, so getting rid of at least one is probably a good idea.

Finally, should we add monthly weighting to the polar and vector plots as well?

brianpm commented on lib/plotting_functions.py:
```diff
@@ -172,6 +183,44 @@ def global_average(fld, wgt, verbose=False):
+        else:
+            spatial_dims = [dimname for dimname in indata.dims if (('lat' in dimname.lower()) or ('lon' in dimname.lower()))]
+
+    if not spatial_dims:
+        warnings.warn("No spatial dimensions were identified, so can not perform average.")
+
+    return weighted.mean(dim=spatial_dims)
```
Looks like `arr.mean(dim=[])` just returns the original `arr` without doing averaging.

nusbaume replied:
Thanks for looking that up @brianpm!  I am worried that if we return the `weighted` array without any dimension reductions from averaging that the scripts using this function could produce un-expected or un-predictable behavior that might be really difficult to track down (e.g. everything "works" but the final values are wrong).  Given this I might go ahead and make the `if not spatial_dims` check and error instead of a warning.  If you or @justin-richling disagree with that plan please let me know.  Thanks!

brianpm replied:
That is perfectly ok with me.

brianpm commented on scripts/analysis/amwg_table.py:
```diff
@@ -376,33 +374,7 @@ def _load_data(dataloc, varname):
-    # we want to average over all non-time dimensions
-    avgdims = [dim for dim in indata.dims if dim != 'time']
-    return weighted.mean(dim=avgdims)
+    return pf.spatial_average(indata)
```
Yeah, I don't see why not. Done.

brianpm commented:
@nusbaume -- I think I addressed all your comments. I also update the polar and vector plots to use `seasonal_mean`. I also added a `.sel(lev=pres)` in the maps for the case when we should be selecting a pressure level. I think that I missed that before. 

I just pushed the updates to my branch -- they appear automatically here, right?

nusbaume reviewed (approved):
Looks good to me!  Once @justin-richling is happy with it we can go ahead and get it merged.

justin-richling reviewed (approved):
Looks good on my end, thanks @brianpm!
